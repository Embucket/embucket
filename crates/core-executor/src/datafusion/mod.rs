//pub mod analyzer;
//pub mod error;
pub mod analyzer;
pub mod error;
pub mod physical_optimizer;
pub mod planner;
pub mod rewriters;
pub mod type_planner;
pub mod visitors;

pub use df_builtins as functions;
