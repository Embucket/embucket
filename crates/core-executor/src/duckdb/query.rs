use crate::error::{self as ex_error, Result};
use arrow_schema::SchemaRef;
use datafusion::arrow::array::RecordBatch;
use datafusion::sql::parser::Statement as DFStatement;
use duckdb::Connection;
use snafu::ResultExt;
use sqlparser::ast::Statement;
use std::sync::Arc;

pub fn execute_duck_db_explain(conn: &Connection, sql: &str) -> Result<Vec<RecordBatch>> {
    let explain_sql = format!("EXPLAIN (format html) {sql}");
    conn.execute("PRAGMA explain_output = 'all'", [])
        .context(ex_error::DuckdbSnafu)?;
    let (res, _) = query_duck_db_arrow(conn, &explain_sql)?;
    Ok(res)
}

pub fn query_duck_db_arrow(
    duckdb_conn: &Connection,
    sql: &str,
) -> Result<(Vec<RecordBatch>, SchemaRef)> {
    // Clone connection for blocking thread
    let sql = sql.to_string();

    // Prepare statement and get schema
    let mut stmt = duckdb_conn
        .prepare(&sql)
        .context(crate::error::DuckdbSnafu)?;
    let result: duckdb::Arrow<'_> = stmt.query_arrow([]).context(crate::error::DuckdbSnafu)?;
    let schema = result.get_schema();
    let res = result.collect();
    Ok((res, schema))
}

#[must_use]
pub fn is_select_statement(stmt: &DFStatement) -> bool {
    matches!(stmt, DFStatement::Statement(inner) if matches!(**inner, Statement::Query(_)))
}

pub fn apply_connection_setup_queries(conn: &Connection, setup_queries: &[Arc<str>]) -> Result<()> {
    for query in setup_queries {
        conn.execute(query, []).context(ex_error::DuckdbSnafu)?;
    }
    Ok(())
}
