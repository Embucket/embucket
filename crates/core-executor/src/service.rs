use std::vec;
use std::{collections::HashMap, sync::Arc};

use bytes::{Buf, Bytes};
use datafusion::arrow::array::RecordBatch;
use datafusion::arrow::csv::ReaderBuilder;
use datafusion::arrow::csv::reader::Format;
use datafusion::catalog::CatalogProvider;
use datafusion::catalog::{MemoryCatalogProvider, MemorySchemaProvider};
use datafusion::datasource::memory::MemTable;
use datafusion_common::TableReference;
use snafu::ResultExt;

use super::error::{self as ex_error, ExecutionError, ExecutionResult};
use super::{models::QueryContext, models::QueryResult, session::UserSession};
use crate::utils::{Config, query_result_to_history};
use core_history::history_store::HistoryStore;
use core_history::store::SlateDBHistoryStore;
use core_metastore::{Metastore, SlateDBMetastore, TableIdent as MetastoreTableIdent};
use core_utils::Db;
use tokio::sync::RwLock;
use uuid::Uuid;

#[async_trait::async_trait]
pub trait ExecutionService: Send + Sync {
    async fn create_session(&self, session_id: String) -> ExecutionResult<Arc<UserSession>>;
    async fn delete_session(&self, session_id: String) -> ExecutionResult<()>;
    async fn query(
        &self,
        session_id: &str,
        query: &str,
        query_context: QueryContext,
    ) -> ExecutionResult<QueryResult>;
    async fn upload_data_to_table(
        &self,
        session_id: &str,
        table_ident: &MetastoreTableIdent,
        data: Bytes,
        file_name: &str,
        format: Format,
    ) -> ExecutionResult<usize>;
}

pub struct CoreExecutionService {
    metastore: Arc<dyn Metastore>,
    history_store: Arc<dyn HistoryStore>,
    df_sessions: Arc<RwLock<HashMap<String, Arc<UserSession>>>>,
    config: Arc<Config>,
}

impl CoreExecutionService {
    pub fn new(
        metastore: Arc<dyn Metastore>,
        history_store: Arc<dyn HistoryStore>,
        config: Arc<Config>,
    ) -> Self {
        Self {
            metastore,
            history_store,
            df_sessions: Arc::new(RwLock::new(HashMap::new())),
            config,
        }
    }
}

#[async_trait::async_trait]
impl ExecutionService for CoreExecutionService {
    #[tracing::instrument(
        name = "ExecutionService::create_session",
        level = "debug",
        skip(self),
        err
    )]
    async fn create_session(&self, session_id: String) -> ExecutionResult<Arc<UserSession>> {
        {
            let sessions = self.df_sessions.read().await;
            if let Some(session) = sessions.get(&session_id) {
                return Ok(session.clone());
            }
        }
        let user_session = Arc::new(
            UserSession::new(
                self.metastore.clone(),
                self.history_store.clone(),
                self.config.clone(),
            )
            .await?,
        );
        {
            tracing::trace!("Acquiring write lock for df_sessions");
            let mut sessions = self.df_sessions.write().await;
            tracing::trace!("Acquired write lock for df_sessions");
            sessions.insert(session_id.clone(), user_session.clone());
        }
        Ok(user_session)
    }

    #[tracing::instrument(
        name = "ExecutionService::delete_session",
        level = "debug",
        skip(self),
        err
    )]
    async fn delete_session(&self, session_id: String) -> ExecutionResult<()> {
        // TODO: Need to have a timeout for the lock
        let mut session_list = self.df_sessions.write().await;
        session_list.remove(&session_id);
        Ok(())
    }

    #[tracing::instrument(name = "ExecutionService::query", level = "debug", skip(self), err)]
    #[allow(clippy::large_futures)]
    async fn query(
        &self,
        session_id: &str,
        query: &str,
        query_context: QueryContext,
    ) -> ExecutionResult<QueryResult> {
        let user_session = {
            let sessions = self.df_sessions.read().await;
            sessions
                .get(session_id)
                .ok_or(ExecutionError::MissingDataFusionSession {
                    id: session_id.to_string(),
                })?
                .clone()
        };

        let mut history_record = self
            .history_store
            .query_record(query, query_context.worksheet_id);
        // Attach the generated query ID to the query context before execution.
        // This ensures consistent tracking and logging of the query across all layers.
        let mut query_obj = user_session.query(
            query,
            query_context.with_query_id(history_record.query_id()),
        );

        let query_result = query_obj.execute().await;
        // Record the query in the sessionâ€™s history, including result count or error message.
        // This ensures all queries are traceable and auditable within a session, which enables
        // features like `last_query_id()` and enhances debugging and observability.
        self.history_store
            .save_query_record(&mut history_record, query_result_to_history(&query_result))
            .await;
        query_result
    }

    #[tracing::instrument(
        name = "ExecutionService::upload_data_to_table",
        level = "debug",
        skip(self, data),
        err,
        ret
    )]
    async fn upload_data_to_table(
        &self,
        session_id: &str,
        table_ident: &MetastoreTableIdent,
        data: Bytes,
        file_name: &str,
        format: Format,
    ) -> ExecutionResult<usize> {
        // TODO: is there a way to avoid temp table approach altogether?
        // File upload works as follows:
        // 1. Convert incoming data to a record batch
        // 2. Create a temporary table in memory
        // 3. Use Execution service to insert data into the target table from the temporary table
        // 4. Drop the temporary table

        // use unique name to support simultaneous uploads
        let unique_id = Uuid::new_v4().to_string().replace('-', "_");
        let user_session = {
            let sessions = self.df_sessions.read().await;
            sessions
                .get(session_id)
                .ok_or(ExecutionError::MissingDataFusionSession {
                    id: session_id.to_string(),
                })?
                .clone()
        };

        let source_table =
            TableReference::full("tmp_db", "tmp_schema", format!("tmp_table_{unique_id}"));
        let target_table = TableReference::full(
            table_ident.database.clone(),
            table_ident.schema.clone(),
            table_ident.table.clone(),
        );
        let inmem_catalog = MemoryCatalogProvider::new();
        inmem_catalog
            .register_schema(
                source_table.schema().unwrap_or_default(),
                Arc::new(MemorySchemaProvider::new()),
            )
            .context(ex_error::DataFusionSnafu)?;
        user_session.ctx.register_catalog(
            source_table.catalog().unwrap_or_default(),
            Arc::new(inmem_catalog),
        );
        // If target table already exists, we need to insert into it
        // otherwise, we need to create it
        let exists = user_session
            .ctx
            .table_exist(target_table.clone())
            .context(ex_error::DataFusionSnafu)?;

        let schema = if exists {
            let table = user_session
                .ctx
                .table(target_table)
                .await
                .context(ex_error::DataFusionSnafu)?;
            table.schema().as_arrow().to_owned()
        } else {
            let (schema, _) = format
                .infer_schema(data.clone().reader(), None)
                .map_err(|e| ExecutionError::DataFusion { source: e.into() })?;
            schema
        };
        let schema = Arc::new(schema);

        // Here we create an arrow CSV reader that infers the schema from the entire dataset
        // (as `None` is passed for the number of rows) and then builds a record batch
        // TODO: This partially duplicates what Datafusion does with `CsvFormat::infer_schema`
        let csv = ReaderBuilder::new(schema.clone())
            .with_format(format)
            .build_buffered(data.reader())
            .context(ex_error::ArrowSnafu)?;

        let batches: Result<Vec<_>, _> = csv.collect();
        let batches = batches.context(ex_error::ArrowSnafu)?;

        let rows_loaded = batches
            .iter()
            .map(|batch: &RecordBatch| batch.num_rows())
            .sum();

        let table = MemTable::try_new(schema, vec![batches]).context(ex_error::DataFusionSnafu)?;
        user_session
            .ctx
            .register_table(source_table.clone(), Arc::new(table))
            .context(ex_error::DataFusionSnafu)?;

        let table = source_table.clone();
        let query = if exists {
            format!("INSERT INTO {table_ident} SELECT * FROM {table}")
        } else {
            format!("CREATE TABLE {table_ident} AS SELECT * FROM {table}")
        };

        let mut query = user_session.query(&query, QueryContext::default());
        Box::pin(query.execute()).await?;

        user_session
            .ctx
            .deregister_table(source_table)
            .context(ex_error::DataFusionSnafu)?;

        Ok(rows_loaded)
    }
}

//Test environment
pub async fn make_text_execution_svc() -> Arc<CoreExecutionService> {
    let db = Db::memory().await;
    let metastore = Arc::new(SlateDBMetastore::new(db.clone()));
    let history_store = Arc::new(SlateDBHistoryStore::new(db));
    Arc::new(CoreExecutionService::new(
        metastore,
        history_store,
        Arc::new(Config::default()),
    ))
}
