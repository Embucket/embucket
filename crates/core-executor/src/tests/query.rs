use crate::models::QueryResultData;
use crate::query::{QueryContext, UserQuery};
use crate::session::{SessionProperty, UserSession};

use crate::error::{ExecutionError, ExecutionResult};
use crate::service::{CoreExecutionService, ExecutionService};
use crate::utils::{Config, DataSerializationFormat};
use core_metastore::Metastore;
use core_metastore::SlateDBMetastore;
use core_metastore::{
    Database as MetastoreDatabase, Schema as MetastoreSchema, SchemaIdent as MetastoreSchemaIdent,
    TableIdent as MetastoreTableIdent, Volume as MetastoreVolume,
};
use datafusion::assert_batches_eq;
use datafusion::sql::parser::{DFParser, Statement as DFStatement};
use datafusion::sql::sqlparser::ast::Statement as SQLStatement;
use datafusion::sql::sqlparser::ast::visit_expressions;
use datafusion::sql::sqlparser::ast::{Expr, ObjectName, ObjectNamePart};
use df_catalog::test_utils::sort_record_batch_by_sortable_columns;
use sqlparser::ast::{
    Function, FunctionArg, FunctionArgExpr, FunctionArgumentList, FunctionArguments, Ident,
};
use sqlparser::ast::{SetExpr, Value};
use std::ops::ControlFlow;
use std::sync::Arc;

struct Test<'a, T> {
    input: &'a str,
    expected: T,
    should_work: bool,
}
impl<'a, T> Test<'a, T> {
    pub const fn new(input: &'a str, expected: T, should_work: bool) -> Self {
        Self {
            input,
            expected,
            should_work,
        }
    }
}
#[tokio::test]
#[allow(
    clippy::unwrap_used,
    clippy::expect_used,
    clippy::explicit_iter_loop,
    clippy::collapsible_match
)]
async fn test_timestamp_keywords_postprocess() {
    let metastore = SlateDBMetastore::new_in_memory().await;
    let session = Arc::new(
        UserSession::new(metastore)
            .await
            .expect("Failed to create user session"),
    );
    let query_context = QueryContext::default();
    let test = vec![
        Test::new(
            "SELECT dateadd(year, 5, '2025-06-01')",
            Value::SingleQuotedString("year".to_owned()),
            true,
        ),
        Test::new(
            "SELECT dateadd(\"year\", 5, '2025-06-01')",
            Value::SingleQuotedString("year".to_owned()),
            true,
        ),
        Test::new(
            "SELECT dateadd('year', 5, '2025-06-01')",
            Value::SingleQuotedString("year".to_owned()),
            true,
        ),
        Test::new(
            "SELECT dateadd(\"'year'\", 5, '2025-06-01')",
            Value::SingleQuotedString("year".to_owned()),
            false,
        ),
        Test::new(
            "SELECT dateadd(\'year\', 5, '2025-06-01')",
            Value::SingleQuotedString("year".to_owned()),
            true,
        ),
        Test::new(
            "SELECT datediff(day, 5, '2025-06-01')",
            Value::SingleQuotedString("day".to_owned()),
            true,
        ),
        Test::new(
            "SELECT datediff('week', 5, '2025-06-01')",
            Value::SingleQuotedString("week".to_owned()),
            true,
        ),
        Test::new(
            "SELECT datediff(nsecond, 10000000, '2025-06-01')",
            Value::SingleQuotedString("nsecond".to_owned()),
            true,
        ),
        Test::new(
            "SELECT date_diff(hour, 5, '2025-06-01')",
            Value::SingleQuotedString("hour".to_owned()),
            true,
        ),
        Test::new(
            "SELECT date_add(us, 100000, '2025-06-01')",
            Value::SingleQuotedString("us".to_owned()),
            true,
        ),
    ];
    for test in test.iter() {
        let query = session.query(test.input, query_context.clone());
        let mut statement = query.parse_query().unwrap();
        UserQuery::postprocess_query_statement(&mut statement);
        if let DFStatement::Statement(statement) = statement {
            visit_expressions(&statement, |expr| {
                if let Expr::Function(Function {
                    name: ObjectName(object_name_parts),
                    args: FunctionArguments::List(FunctionArgumentList { args, .. }),
                    ..
                }) = expr
                {
                    match object_name_parts.first().unwrap() {
                        ObjectNamePart::Identifier(ident) => match ident.value.as_str() {
                            "dateadd" | "date_add" | "datediff" | "date_diff" => {
                                if let FunctionArg::Unnamed(FunctionArgExpr::Expr(ident)) =
                                    args.iter().next().unwrap()
                                {
                                    if let Expr::Value(found) = ident {
                                        if test.should_work {
                                            assert_eq!(
                                                found.to_string(),
                                                test.expected.to_string()
                                            );
                                        } else {
                                            assert_ne!(
                                                found.to_string(),
                                                test.expected.to_string()
                                            );
                                        }
                                    }
                                }
                            }
                            _ => {}
                        },
                    }
                }
                ControlFlow::<()>::Continue(())
            });
        }
    }
}

#[allow(clippy::unwrap_used)]
#[test]
fn test_postprocess_query_statement_functions_expressions() {
    let args: [(&str, &str); 14] = [
        ("select year(ts)", "SELECT date_part('year', ts)"),
        ("select dayofyear(ts)", "SELECT date_part('doy', ts)"),
        ("select day(ts)", "SELECT date_part('day', ts)"),
        ("select dayofmonth(ts)", "SELECT date_part('day', ts)"),
        ("select dayofweek(ts)", "SELECT date_part('dow', ts)"),
        ("select month(ts)", "SELECT date_part('month', ts)"),
        ("select weekofyear(ts)", "SELECT date_part('week', ts)"),
        ("select week(ts)", "SELECT date_part('week', ts)"),
        ("select hour(ts)", "SELECT date_part('hour', ts)"),
        ("select minute(ts)", "SELECT date_part('minute', ts)"),
        ("select second(ts)", "SELECT date_part('second', ts)"),
        ("select minute(ts)", "SELECT date_part('minute', ts)"),
        // Do nothing
        ("select yearofweek(ts)", "SELECT yearofweek(ts)"),
        ("select yearofweekiso(ts)", "SELECT yearofweekiso(ts)"),
    ];

    for (init, exp) in args {
        let statement = DFParser::parse_sql(init).unwrap().pop_front();
        if let Some(mut s) = statement {
            UserQuery::postprocess_query_statement(&mut s);
            assert_eq!(s.to_string(), exp);
        }
    }
}

#[tokio::test]
#[allow(clippy::expect_used, clippy::manual_let_else, clippy::too_many_lines)]
async fn test_context_name_injection() {
    let metastore = SlateDBMetastore::new_in_memory().await;
    let session = Arc::new(
        UserSession::new(metastore)
            .await
            .expect("Failed to create user session"),
    );
    let query1 = session.query("SELECT * FROM table1", QueryContext::default());
    let query_statement = if let DFStatement::Statement(statement) =
        query1.parse_query().expect("Failed to parse query")
    {
        if let SQLStatement::Query(query) = *statement {
            query
        } else {
            panic!("Failed to parse query");
        }
    } else {
        panic!("Failed to parse query");
    };
    let select_statement1 = if let SetExpr::Select(select) = *query_statement.body {
        select
    } else {
        panic!("Failed to parse query");
    };
    let from1 = select_statement1.from;

    let query2 = session.query(
        "SELECT * from table2",
        QueryContext::new(Some("db2".to_string()), Some("sch2".to_string()), None),
    );
    let query_statement2 = if let DFStatement::Statement(statement) =
        query2.parse_query().expect("Failed to parse query")
    {
        if let SQLStatement::Query(query) = *statement {
            query
        } else {
            panic!("Failed to parse query");
        }
    } else {
        panic!("Failed to parse query");
    };
    let select_statement2 = if let SetExpr::Select(select) = *query_statement2.body {
        select
    } else {
        panic!("Failed to parse query");
    };
    let from2 = select_statement2.from;

    // This will also test the default public schema
    session
        .set_session_variable(
            true,
            vec![(
                "catalog".to_string(),
                SessionProperty::from_str("db3".to_string()),
            )]
            .into_iter()
            .collect(),
        )
        .expect("Failed to set session variable");
    let query3 = session.query("SELECT * from table3", QueryContext::default());
    let query_statement3 = if let DFStatement::Statement(statement) =
        query3.parse_query().expect("Failed to parse query")
    {
        if let SQLStatement::Query(query) = *statement {
            query
        } else {
            panic!("Failed to parse query");
        }
    } else {
        panic!("Failed to parse query");
    };
    let select_statement3 = if let SetExpr::Select(select) = *query_statement3.body {
        select
    } else {
        panic!("Failed to parse query");
    };
    let from3 = select_statement3.from;

    // Test
    session
        .set_session_variable(
            true,
            vec![
                (
                    "catalog".to_string(),
                    SessionProperty::from_str("db4".to_string()),
                ),
                (
                    "schema".to_string(),
                    SessionProperty::from_str("sch4".to_string()),
                ),
            ]
            .into_iter()
            .collect(),
        )
        .expect("Failed to set session variable");
    let query4 = session.query(
        "SELECT * from table4 INNER JOIN table4_1 ON 1=1",
        QueryContext::default(),
    );
    let query_statement4 = if let DFStatement::Statement(statement) =
        query4.parse_query().expect("Failed to parse query")
    {
        if let SQLStatement::Query(query) = *statement {
            query
        } else {
            panic!("Failed to parse query");
        }
    } else {
        panic!("Failed to parse query");
    };
    let select_statement4 = if let SetExpr::Select(select) = *query_statement4.body {
        select
    } else {
        panic!("Failed to parse query");
    };
    let from4 = select_statement4.from;
    insta::assert_debug_snapshot!((from1, from2, from3, from4));
}

#[tokio::test]
async fn test_create_table_with_timestamp_nanosecond() {
    let (execution_svc, _, session_id) = prepare_env().await;
    let table_ident = MetastoreTableIdent {
        database: "embucket".to_string(),
        schema: "public".to_string(),
        table: "target_table".to_string(),
    };
    // Verify that the file was uploaded successfully by running select * from the table
    let query = format!(
        "CREATE TABLE {}.{}.{} (id INT, ts TIMESTAMP_NTZ(9)) as VALUES (1, '2025-04-09T21:11:23'), (2, '2025-04-09T21:11:00');",
        table_ident.database, table_ident.schema, table_ident.table
    );
    let QueryResultData { records: rows, .. } = execution_svc
        .query(&session_id, &query, QueryContext::default())
        .await
        .expect("Failed to execute query");

    assert_batches_eq!(
        &[
            "+-------+",
            "| count |",
            "+-------+",
            "| 2     |",
            "+-------+",
        ],
        &rows
    );
}

#[tokio::test]
async fn test_drop_table() {
    let (execution_svc, _, session_id) = prepare_env().await;
    let table_ident = MetastoreTableIdent {
        database: "embucket".to_string(),
        schema: "public".to_string(),
        table: "target_table".to_string(),
    };
    // Verify that the file was uploaded successfully by running select * from the table
    let query = format!("CREATE TABLE {table_ident} (id INT) as VALUES (1), (2);");
    let QueryResultData { records: rows, .. } = execution_svc
        .query(&session_id, &query, QueryContext::default())
        .await
        .expect("Failed to execute query");

    assert_batches_eq!(
        &[
            "+-------+",
            "| count |",
            "+-------+",
            "| 2     |",
            "+-------+",
        ],
        &rows
    );

    let query = format!("DROP TABLE {table_ident};");
    execution_svc
        .query(&session_id, &query, QueryContext::default())
        .await
        .expect("Failed to execute query");

    // Verify that the table is not exists
    let query = format!("SELECT * FROM {table_ident};");
    let res = execution_svc
        .query(&session_id, &query, QueryContext::default())
        .await;

    assert!(res.is_err());
    if let Err(err) = res {
        assert_eq!(
            err.to_string(),
            format!("DataFusion error: Error during planning: table '{table_ident}' not found")
        );
    }
}

#[tokio::test]
async fn test_create_schema() {
    let (execution_svc, metastore, session_id) = prepare_env().await;
    let schema_ident = MetastoreSchemaIdent {
        database: "embucket".to_string(),
        schema: "public_new".to_string(),
    };
    let query = format!("CREATE SCHEMA {schema_ident};");
    execution_svc
        .query(&session_id, &query, QueryContext::default())
        .await
        .expect("Failed to execute query");
    // TODO use "SHOW SCHEMAS" sql
    metastore
        .get_schema(&schema_ident)
        .await
        .expect("Failed to get schema");
}

#[tokio::test]
#[allow(clippy::unwrap_used)]
async fn test_resolve_table_object_name() {
    let metastore = SlateDBMetastore::new_in_memory().await;
    let session = UserSession::new(metastore)
        .await
        .expect("Failed to create session");
    let query = UserQuery::new(Arc::from(session), "", QueryContext::default());

    let test_cases: [(Vec<ObjectNamePart>, ExecutionResult<String>); 4] = [
        (
            vec![ObjectNamePart::Identifier(Ident::new("table"))],
            Ok("embucket.public.table".to_string()),
        ),
        (
            vec![
                ObjectNamePart::Identifier(Ident::new("test_schema")),
                ObjectNamePart::Identifier(Ident::new("table")),
            ],
            Ok("embucket.test_schema.table".to_string()),
        ),
        (
            vec![
                ObjectNamePart::Identifier(Ident::new("test_db")),
                ObjectNamePart::Identifier(Ident::new("test_schema")),
                ObjectNamePart::Identifier(Ident::new("table")),
            ],
            Ok("test_db.test_schema.table".to_string()),
        ),
        (
            vec![
                ObjectNamePart::Identifier(Ident::new("test_db")),
                ObjectNamePart::Identifier(Ident::new("test_schema")),
                ObjectNamePart::Identifier(Ident::new("table")),
                ObjectNamePart::Identifier(Ident::new("col")),
            ],
            Err(ExecutionError::InvalidTableIdentifier {
                ident: "test_db.test_schema.table.col".to_string(),
            }),
        ),
    ];
    for (test_case, expected) in test_cases {
        let result = query.resolve_table_object_name(test_case.clone());
        if result.is_err() {
            assert_eq!(
                result.err().unwrap().to_string(),
                expected.err().unwrap().to_string()
            );
        } else {
            assert_eq!(result.unwrap().to_string(), expected.unwrap());
        }
    }
}

#[tokio::test]
#[allow(clippy::unwrap_used)]
async fn test_resolve_schema_object_name() {
    let metastore = SlateDBMetastore::new_in_memory().await;
    let session = UserSession::new(metastore)
        .await
        .expect("Failed to create session");
    let query = UserQuery::new(Arc::from(session), "", QueryContext::default());

    let test_cases: [(Vec<ObjectNamePart>, ExecutionResult<String>); 3] = [
        (
            vec![ObjectNamePart::Identifier(Ident::new("schema"))],
            Ok("embucket.schema".to_string()),
        ),
        (
            vec![
                ObjectNamePart::Identifier(Ident::new("test_db")),
                ObjectNamePart::Identifier(Ident::new("schema")),
            ],
            Ok("test_db.schema".to_string()),
        ),
        (
            vec![
                ObjectNamePart::Identifier(Ident::new("test_db")),
                ObjectNamePart::Identifier(Ident::new("test_schema")),
                ObjectNamePart::Identifier(Ident::new("table")),
            ],
            Err(ExecutionError::InvalidSchemaIdentifier {
                ident: "test_db.test_schema.table".to_string(),
            }),
        ),
    ];
    for (test_case, expected) in test_cases {
        let res = query.resolve_schema_object_name(test_case.clone());
        if res.is_err() {
            assert_eq!(
                res.err().unwrap().to_string(),
                expected.err().unwrap().to_string()
            );
        } else {
            assert_eq!(res.unwrap().to_string(), expected.unwrap());
        }
    }
}

async fn prepare_env() -> (CoreExecutionService, Arc<SlateDBMetastore>, String) {
    let metastore = SlateDBMetastore::new_in_memory().await;
    metastore
        .create_volume(
            &"test_volume".to_string(),
            MetastoreVolume::new(
                "test_volume".to_string(),
                core_metastore::VolumeType::Memory,
            ),
        )
        .await
        .expect("Failed to create volume");
    metastore
        .create_database(
            &"embucket".to_string(),
            MetastoreDatabase {
                ident: "embucket".to_string(),
                properties: None,
                volume: "test_volume".to_string(),
            },
        )
        .await
        .expect("Failed to create database");
    let schema_ident = MetastoreSchemaIdent {
        database: "embucket".to_string(),
        schema: "public".to_string(),
    };
    metastore
        .create_schema(
            &schema_ident.clone(),
            MetastoreSchema {
                ident: schema_ident,
                properties: None,
            },
        )
        .await
        .expect("Failed to create schema");

    let execution_svc = CoreExecutionService::new(
        metastore.clone(),
        Config {
            dbt_serialization_format: DataSerializationFormat::Json,
        },
    );

    let session_id = "test_session_id";
    execution_svc
        .create_session(session_id.to_string())
        .await
        .expect("Failed to create session");
    (execution_svc, metastore, session_id.to_string())
}

static TABLE_SETUP: &str = include_str!(r"./table_setup.sql");

#[allow(clippy::unwrap_used, clippy::expect_used)]
pub async fn create_df_session() -> Arc<UserSession> {
    let metastore = SlateDBMetastore::new_in_memory().await;
    metastore
        .create_volume(
            &"test_volume".to_string(),
            MetastoreVolume::new(
                "test_volume".to_string(),
                core_metastore::VolumeType::Memory,
            ),
        )
        .await
        .expect("Failed to create volume");
    metastore
        .create_database(
            &"embucket".to_string(),
            MetastoreDatabase {
                ident: "embucket".to_string(),
                properties: None,
                volume: "test_volume".to_string(),
            },
        )
        .await
        .expect("Failed to create database");
    let schema_ident = MetastoreSchemaIdent {
        database: "embucket".to_string(),
        schema: "public".to_string(),
    };
    metastore
        .create_schema(
            &schema_ident.clone(),
            MetastoreSchema {
                ident: schema_ident,
                properties: None,
            },
        )
        .await
        .expect("Failed to create schema");

    let user_session = Arc::new(
        UserSession::new(metastore)
            .await
            .expect("Failed to create user session"),
    );

    for query in TABLE_SETUP.split(';') {
        if !query.is_empty() {
            let mut query = user_session.query(query, QueryContext::default());
            query.execute().await.unwrap();
            //ctx.sql(query).await.unwrap().collect().await.unwrap();
        }
    }
    user_session
}

#[macro_export]
macro_rules! test_query {
    (
        $test_fn_name:ident,
        $query:expr
        $(, pre_queries = [$($pre_queries:expr),* $(,)?])?
        $(, sort_all = $sort_all:expr)?
    ) => {
        paste::paste! {
            #[tokio::test]
            async fn [< query_ $test_fn_name >]() {
                let ctx = create_df_session().await;

                // Execute all pre-queries (if provided) to set up the session context
                $(
                    $(
                        {
                            let mut q = ctx.query($pre_queries, crate::query::QueryContext::default());
                            q.execute().await.unwrap();
                        }
                    )*
                )?

                let mut query = ctx.query($query, crate::query::QueryContext::default());
                let res = query.execute().await;
                let sort_all = false $(|| $sort_all)?;

                insta::with_settings!({
                    description => stringify!($query),
                    omit_expression => true,
                    prepend_module_to_snapshot => false
                }, {
                    let df = match res {
                        Ok(mut record_batches) => {
                            if sort_all {
                                for batch in &mut record_batches {
                                    *batch = sort_record_batch_by_sortable_columns(batch);
                                }
                            }
                            Ok(datafusion::arrow::util::pretty::pretty_format_batches(&record_batches).unwrap().to_string())
                        },
                        Err(e) => Err(format!("Error: {e}"))
                    };

                    let df = df.map(|df| df.split('\n').map(|s| s.to_string()).collect::<Vec<String>>());
                    insta::assert_debug_snapshot!((df));
                });
            }
        }
    };
}

test_query!(select_date_add_diff, "SELECT dateadd(day, 5, '2025-06-01')");
test_query!(func_date_add, "SELECT date_add(day, 30, '2025-01-06')");
// // SELECT
test_query!(select_star, "SELECT * FROM employee_table");
// FIXME: ILIKE is not supported yet
// test_query!(select_ilike, "SELECT * ILIKE '%id%' FROM employee_table;");
test_query!(
    select_exclude,
    "SELECT * EXCLUDE department_id FROM employee_table;"
);
test_query!(
    select_exclude_multiple,
    "SELECT * EXCLUDE (department_id, employee_id) FROM employee_table;"
);

test_query!(
    qualify,
    "SELECT product_id, retail_price, quantity, city
    FROM sales
    QUALIFY ROW_NUMBER() OVER (PARTITION BY city ORDER BY retail_price) = 1;"
);

// SHOW DATABASES
test_query!(show_databases, "SHOW DATABASES", sort_all = true);

// SHOW SCHEMAS
test_query!(show_schemas, "SHOW SCHEMAS", sort_all = true);
test_query!(
    show_schemas_starts_with,
    "SHOW SCHEMAS STARTS WITH 'publ'",
    sort_all = true
);
test_query!(
    show_schemas_in_db,
    "SHOW SCHEMAS IN embucket",
    sort_all = true
);
test_query!(
    show_schemas_in_db_and_prefix,
    "SHOW SCHEMAS IN embucket STARTS WITH 'pub'",
    sort_all = true
);

// SHOW TABLES
test_query!(show_tables, "SHOW TABLES", sort_all = true);
test_query!(
    show_tables_starts_with,
    "SHOW TABLES STARTS WITH 'dep'",
    sort_all = true
);
test_query!(
    show_tables_in_schema,
    "SHOW TABLES IN public",
    sort_all = true
);
test_query!(
    show_tables_in_schema_full,
    "SHOW TABLES IN embucket.public",
    sort_all = true
);
test_query!(
    show_tables_in_schema_and_prefix,
    "SHOW TABLES IN public STARTS WITH 'dep'",
    sort_all = true
);

// SHOW VIEWS
test_query!(show_views, "SHOW VIEWS", sort_all = true);
test_query!(
    show_views_starts_with,
    "SHOW VIEWS STARTS WITH 'schem'",
    sort_all = true
);
test_query!(
    show_views_in_schema,
    "SHOW VIEWS IN information_schema",
    sort_all = true
);
test_query!(
    show_views_in_schema_full,
    "SHOW VIEWS IN embucket.information_schema",
    sort_all = true
);
test_query!(
    show_views_in_schema_and_prefix,
    "SHOW VIEWS IN information_schema STARTS WITH 'schem'",
    sort_all = true
);

// SHOW COLUMNS
test_query!(show_columns, "SHOW COLUMNS", sort_all = true);
test_query!(
    show_columns_in_table,
    "SHOW COLUMNS IN employee_table",
    sort_all = true
);
test_query!(
    show_columns_in_table_full,
    "SHOW COLUMNS IN embucket.public.employee_table",
    sort_all = true
);
test_query!(
    show_columns_starts_with,
    "SHOW COLUMNS IN employee_table STARTS WITH 'last_'",
    sort_all = true
);

// SHOW OBJECTS
test_query!(show_objects, "SHOW OBJECTS", sort_all = true);
test_query!(
    show_objects_starts_with,
    "SHOW OBJECTS STARTS WITH 'dep'",
    sort_all = true
);
test_query!(
    show_objects_in_schema,
    "SHOW OBJECTS IN public",
    sort_all = true
);
test_query!(
    show_objects_in_schema_full,
    "SHOW OBJECTS IN embucket.public",
    sort_all = true
);
test_query!(
    show_objects_in_schema_and_prefix,
    "SHOW OBJECTS IN public STARTS WITH 'dep'",
    sort_all = true
);

// SHOW VARIABLES
test_query!(
    show_variables_multiple,
    "SHOW VARIABLES",
    pre_queries = ["SET v1 = 'test'", "SET v2 = 1", "SET v3 = true"],
    sort_all = true
);
