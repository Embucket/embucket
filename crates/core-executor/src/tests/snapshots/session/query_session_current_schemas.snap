---
source: crates/core-executor/src/tests/query.rs
description: "\"SELECT CURRENT_SCHEMAS()\""
---
Ok(
    [
        "+-------------------+",
        "| current_schemas   |",
        "+-------------------+",
        "| [embucket.public] |",
        "+-------------------+",
    ],
)
